File: lib/mongodb/bangumiRepository.ts
import {getBangumiCollection} from './connection'
import {type BangumiCollection, type BangumiImages, type BangumiRating, ROLE_TYPE_NAMES, SUBJECT_TYPE_NAMES, WEEKDAY_NAMES,} from '@/lib/bangumi'

// Bangumi 条目类型定义 (基于 MONGODB_SCHEMA.md)
export interface BangumiSubjectDoc {
    _id: number
    name: string
    name_cn: string
    type: number
    type_info?: {
        id: number
        name: string
    }
    platform?: number
    platform_info?: {
        id: number
        type: string
        type_cn: string
        alias: string
        wiki_tpl: string
    }
    infobox?: {
        type: string
        type_cn: string
        fields: Record<string, any>
    }
    summary: string
    nsfw: boolean
    meta_tags: string[]
    series: boolean
    date: string
    meta?: {
        score: number
        score_details: Record<string, number>
        rank: number
        tags: string[]
        favorite: {
            wish: number
            done: number
            doing: number
            on_hold: number
            dropped: number
        }
    }
    _raw?: Record<string, any>
    _sync?: {
        created_at: string
        updated_at: string
        version: number
        source_file: string
    }
}

// Bangumi 人物类型定义
export interface BangumiPersonDoc {
    _id: number
    type: number
    name: string
    infobox?: {
        type: string
        type_cn: string
        fields: Record<string, any>
    }
    summary: string
    career: string[]
    collects: number
    comments: number
    _sync?: {
        created_at: string
        updated_at: string
        version: number
        source_file: string
    }
}

// Bangumi 角色类型定义 (MongoDB 文档)
// noinspection JSUnusedGlobalSymbols
export interface BangumiCharacterDoc {
    _id: number
    name: string
    infobox?: {
        type: string
        type_cn: string
        fields: Record<string, any>
    }
    summary: string
    role: number
    collects: number
    comments: number
    _sync?: {
        created_at: string
        updated_at: string
        version: number
        source_file: string
    }
}

// Bangumi 剧集类型定义
export interface BangumiEpisodeDoc {
    _id: number
    subject_id: number
    type: number
    name: string
    name_cn: string
    sort: number
    airdate: string
    duration: string
    description: string
    disc: number
    _sync?: {
        created_at: string
        updated_at: string
        version: number
        source_file: string
    }
}

// 条目与人物关联
export interface BangumiSubjectPersonDoc {
    _id: string
    subject_id: number
    person_id: number
    position: number
    position_info?: {
        id: number
        cn: string
        en: string
        jp: string
        categories: string[]
    }
    appear_eps: string
    _sync?: {
        created_at: string
        updated_at: string
        version: number
        source_file: string
    }
}

// 条目与角色关联
export interface BangumiSubjectCharacterDoc {
    _id: string
    subject_id: number
    character_id: number
    type: number
    order: number
    _sync?: {
        created_at: string
        updated_at: string
        version: number
        source_file: string
    }
}

// 人物与角色关联 (配音)
// noinspection JSUnusedGlobalSymbols
export interface BangumiPersonCharacterDoc {
    _id: string
    person_id: number
    character_id: number
    subject_id: number
    summary: string
    _sync?: {
        created_at: string
        updated_at: string
        version: number
        source_file: string
    }
}

// 条目关系
export interface BangumiSubjectRelationDoc {
    _id: string
    subject_id: number
    related_subject_id: number
    relation_type: number
    relation_info?: {
        id: number
        cn: string
        en: string
        desc: string
    }
    order: number
    _sync?: {
        created_at: string
        updated_at: string
        version: number
        source_file: string
    }
}

// 重新导出共享类型
export type {BangumiImages, BangumiRating, BangumiCollection}

// 条目详情 (含关联数据)
export interface BangumiSubjectDetail extends BangumiSubjectDoc {
    staff?: BangumiStaffItem[]
    characters?: BangumiCharacterItem[]
    episodes?: BangumiEpisodeDoc[]
    relations?: BangumiSubjectRelationItem[]
}

export interface BangumiStaffItem {
    person_id: number
    name: string
    name_cn?: string
    position: string
    url: string
}

export interface BangumiCharacterItem {
    character_id: number
    name: string
    name_cn?: string
    role_type: number
    order: number
    url: string
}

export interface BangumiSubjectRelationItem {
    subject_id: number
    name: string
    name_cn?: string
    relation_type: string
    url: string
}

// 重新导出常量
export {SUBJECT_TYPE_NAMES, WEEKDAY_NAMES, ROLE_TYPE_NAMES}

// ============ 集合获取函数 ============

export function getSubjectsCollection() {
    return getBangumiCollection<BangumiSubjectDoc>('subjects')
}

export function getPersonsCollection() {
    return getBangumiCollection<BangumiPersonDoc>('persons')
}

export function getEpisodesCollection() {
    return getBangumiCollection<BangumiEpisodeDoc>('episodes')
}

export function getSubjectPersonsCollection() {
    return getBangumiCollection<BangumiSubjectPersonDoc>('subject_persons')
}

export function getSubjectCharactersCollection() {
    return getBangumiCollection<BangumiSubjectCharacterDoc>('subject_characters')
}

export function getSubjectRelationsCollection() {
    return getBangumiCollection<BangumiSubjectRelationDoc>('subject_relations')
}

// ============ 查询函数 ============

/**
 * 根据 ID 获取条目详情
 */
export async function getSubjectById(subjectId: number): Promise<BangumiSubjectDoc | null> {
    try {
        const collection = getSubjectsCollection()
        return await collection.findOne({_id: subjectId})
    } catch (error) {
        console.error('[bangumi] getSubjectById error:', error)
        return null
    }
}

/**
 * 获取条目的制作人员
 */
export async function getSubjectStaff(subjectId: number): Promise<BangumiStaffItem[]> {
    try {
        const collection = getSubjectPersonsCollection()
        getPersonsCollection();
        const pipeline = [
            {$match: {subject_id: subjectId}},
            {
                $lookup: {
                    from: 'persons',
                    localField: 'person_id',
                    foreignField: '_id',
                    as: 'person',
                },
            },
            {$unwind: '$person'},
            {
                $project: {
                    person_id: '$person._id',
                    name: '$person.name',
                    name_cn: '$person.infobox.fields.简体中文名',
                    position: '$position_info.cn',
                    url: {$concat: ['https://bgm.tv/person/', {$toString: '$person._id'}]},
                },
            },
        ]

        const results = await collection.aggregate(pipeline).toArray()
        return results as BangumiStaffItem[]
    } catch (error) {
        console.error('[bangumi] getSubjectStaff error:', error)
        return []
    }
}

/**
 * 获取条目的角色
 */
export async function getSubjectCharacters(subjectId: number): Promise<BangumiCharacterItem[]> {
    try {
        const collection = getSubjectCharactersCollection()

        const pipeline = [
            {$match: {subject_id: subjectId}},
            {
                $lookup: {
                    from: 'characters',
                    localField: 'character_id',
                    foreignField: '_id',
                    as: 'character',
                },
            },
            {$unwind: '$character'},
            {
                $project: {
                    character_id: '$character._id',
                    name: '$character.name',
                    name_cn: '$character.infobox.fields.简体中文名',
                    role_type: '$type',
                    order: '$order',
                    url: {$concat: ['https://bgm.tv/character/', {$toString: '$character._id'}]},
                },
            },
            {$sort: {order: 1}},
        ]

        const results = await collection.aggregate(pipeline).toArray()
        return results as BangumiCharacterItem[]
    } catch (error) {
        console.error('[bangumi] getSubjectCharacters error:', error)
        return []
    }
}

/**
 * 获取条目的剧集
 */
export async function getSubjectEpisodes(subjectId: number): Promise<BangumiEpisodeDoc[]> {
    try {
        const collection = getEpisodesCollection()
        return await collection
            .find({subject_id: subjectId})
            .sort({sort: 1})
            .toArray()
    } catch (error) {
        console.error('[bangumi] getSubjectEpisodes error:', error)
        return []
    }
}

/**
 * 获取条目的关联条目
 */
export async function getSubjectRelations(subjectId: number): Promise<BangumiSubjectRelationItem[]> {
    try {
        const collection = getSubjectRelationsCollection()

        const pipeline = [
            {$match: {subject_id: subjectId}},
            {
                $lookup: {
                    from: 'subjects',
                    localField: 'related_subject_id',
                    foreignField: '_id',
                    as: 'related_subject',
                },
            },
            {$unwind: '$related_subject'},
            {
                $project: {
                    subject_id: '$related_subject._id',
                    name: '$related_subject.name',
                    name_cn: '$related_subject.name_cn',
                    relation_type: '$relation_info.cn',
                    url: {$concat: ['https://bgm.tv/subject/', {$toString: '$related_subject._id'}]},
                },
            },
        ]

        const results = await collection.aggregate(pipeline).toArray()
        return results as BangumiSubjectRelationItem[]
    } catch (error) {
        console.error('[bangumi] getSubjectRelations error:', error)
        return []
    }
}

/**
 * 获取完整条目详情 (含关联数据)
 */
export async function getSubjectDetail(subjectId: number): Promise<BangumiSubjectDetail | null> {
    const subject = await getSubjectById(subjectId)
    if (!subject) return null

    const [staff, characters, episodes, relations] = await Promise.all([
        getSubjectStaff(subjectId),
        getSubjectCharacters(subjectId),
        getSubjectEpisodes(subjectId),
        getSubjectRelations(subjectId),
    ])

    return {
        ...subject,
        staff,
        characters,
        episodes,
        relations,
    }
}

/**
 * 获取所有条目 (用于同步到 Meilisearch)
 */
export async function getAllSubjects(
    options: {
        batchSize?: number
        skip?: number
    } = {}
): Promise<BangumiSubjectDoc[]> {
    try {
        const collection = getSubjectsCollection()
        const {batchSize = 1000, skip = 0} = options

        return await collection
            .find({})
            .skip(skip)
            .limit(batchSize)
            .toArray()
    } catch (error) {
        console.error('[bangumi] getAllSubjects error:', error)
        return []
    }
}

/**
 * 获取条目总数
 */
export async function getTotalSubjectsCount(): Promise<number> {
    try {
        const collection = getSubjectsCollection()
        return await collection.countDocuments({})
    } catch (error) {
        console.error('[bangumi] getTotalSubjectsCount error:', error)
        return 0
    }
}

File: src/db/bangumirepository.cpp
#include "db/bangumirepository.h"

#include "db/connection.h"
#include <mongocxx/collection.hpp>
#include <mongocxx/pipeline.hpp>
#include <bsoncxx/builder/basic/document.hpp>
#include <bsoncxx/builder/basic/kvp.hpp>
#include <bsoncxx/builder/basic/array.hpp>

#include "bsonutils.h"

using bsoncxx::builder::basic::kvp;
using bsoncxx::builder::basic::make_document;
using bsoncxx::builder::basic::make_array;

static BangumiSubjectDoc parseSubjectDoc(const bsoncxx::document::view &view) {
    BangumiSubjectDoc s;
    if (view["_id"]) s.id = BsonUtils::toInt32(view["_id"]);
    if (view["name"]) s.name = BsonUtils::toQString(view["name"]);
    if (view["name_cn"]) s.nameCn = BsonUtils::toQString(view["name_cn"]);
    if (view["type"]) s.type = BsonUtils::toInt32(view["type"]);
    if (view["summary"]) s.summary = BsonUtils::toQString(view["summary"]);
    if (view["nsfw"] && view["nsfw"].type() == bsoncxx::type::k_bool)
        s.nsfw = view["nsfw"].get_bool().value;
    if (view["date"]) s.date = BsonUtils::toQString(view["date"]);
    if (view["meta"] && view["meta"].type() == bsoncxx::type::k_document) {
        const auto meta = view["meta"].get_document().value;
        if (meta["score"]) s.score = BsonUtils::toReal(meta["score"]);
        if (meta["rank"]) s.rank = BsonUtils::toInt32(meta["rank"]);
        if (meta["score_details"] && meta["score_details"].type() == bsoncxx::type::k_document) {
            for (auto &&kv: meta["score_details"].get_document().value) {
                s.scoreDetails.insert(BsonUtils::toQString(kv.key()), BsonUtils::toInt32(kv));
            }
        }
        if (meta["favorite"] && meta["favorite"].type() == bsoncxx::type::k_document) {
            auto fav = meta["favorite"].get_document().value;
            if (fav["wish"]) s.wish = BsonUtils::toInt32(fav["wish"]);
            if (fav["done"]) s.collect = BsonUtils::toInt32(fav["done"]);
            if (fav["doing"]) s.doing = BsonUtils::toInt32(fav["doing"]);
            if (fav["on_hold"]) s.onHold = BsonUtils::toInt32(fav["on_hold"]);
            if (fav["dropped"]) s.dropped = BsonUtils::toInt32(fav["dropped"]);
        }
    }
    return s;
}

DbResult<BangumiSubjectDoc> BangumiRepository::getSubjectById(qint32 subjectId) {
    try {
        if (!MongoConnection::instance().isConnected()) return BangumiSubjectDoc{};
        auto db = MongoConnection::instance().database("bangumi");
        auto coll = db["subjects"];
        auto filter = make_document(kvp("_id", subjectId));
        if (auto result = coll.find_one(filter.view()); result) {
            return parseSubjectDoc(result->view());
        }
        return BangumiSubjectDoc{};
    } catch (const std::exception &e) {
        return std::unexpected(QStringLiteral("getSubjectById failed: ") + QString::fromUtf8(e.what()));
    }
}

DbResult<QList<BangumiStaffItem> > BangumiRepository::getSubjectStaff(qint32 subjectId) {
    try {
        if (!MongoConnection::instance().isConnected()) return QList<BangumiStaffItem>{};
        auto db = MongoConnection::instance().database("bangumi");
        auto coll = db["subject_persons"];

        mongocxx::pipeline pipeline;
        pipeline.match(make_document(kvp("subject_id", subjectId)));
        pipeline.lookup(make_document(
            kvp("from", "persons"),
            kvp("localField", "person_id"),
            kvp("foreignField", "_id"),
            kvp("as", "person")
        ));
        pipeline.unwind(make_document(kvp("path", "$person")));
        pipeline.project(make_document(
            kvp("person_id", "$person._id"),
            kvp("name", "$person.name"),
            kvp("name_cn", "$person.infobox.fields.简体中文名"),
            kvp("position", "$position_info.cn"),
            kvp("url", make_document(kvp("$concat", make_array(QStringLiteral("https://bgm.tv/person/").toStdString(), make_document(kvp("$toString", "$person._id"))))))
        ));

        auto cursor = coll.aggregate(pipeline);
        QList<BangumiStaffItem> list;
        for (auto &&doc: cursor) {
            BangumiStaffItem item;
            if (auto elem = doc["person_id"]; elem) item.personId = BsonUtils::toInt32(elem);
            if (auto elem = doc["name"]; elem) item.name = BsonUtils::toQString(elem);
            if (auto elem = doc["name_cn"]; elem) item.nameCn = BsonUtils::toQString(elem);
            if (auto elem = doc["position"]; elem) item.position = BsonUtils::toQString(elem);
            if (auto elem = doc["url"]; elem) item.url = BsonUtils::toQString(elem);
            list.push_back(item);
        }
        return list;
    } catch (const std::exception &e) {
        return std::unexpected(QStringLiteral("getSubjectStaff failed: ") + QString::fromUtf8(e.what()));
    }
}

DbResult<QList<BangumiCharacterItem> > BangumiRepository::getSubjectCharacters(qint32 subjectId) {
    try {
        if (!MongoConnection::instance().isConnected()) return QList<BangumiCharacterItem>{};
        auto db = MongoConnection::instance().database("bangumi");
        auto coll = db["subject_characters"];

        mongocxx::pipeline pipeline;
        pipeline.match(make_document(kvp("subject_id", subjectId)));
        pipeline.lookup(make_document(
            kvp("from", "characters"),
            kvp("localField", "character_id"),
            kvp("foreignField", "_id"),
            kvp("as", "character")
        ));
        pipeline.unwind(make_document(kvp("path", "$character")));
        pipeline.project(make_document(
            kvp("character_id", "$character._id"),
            kvp("name", "$character.name"),
            kvp("name_cn", "$character.infobox.fields.简体中文名"),
            kvp("role_type", "$type"),
            kvp("order", "$order"),
            kvp("url", make_document(kvp("$concat", make_array(QStringLiteral("https://bgm.tv/character/").toStdString(), make_document(kvp("$toString", "$character._id"))))))
        ));
        pipeline.sort(make_document(kvp("order", 1)));

        auto cursor = coll.aggregate(pipeline);
        QList<BangumiCharacterItem> list;
        for (auto &&doc: cursor) {
            BangumiCharacterItem item;
            if (auto elem = doc["character_id"]; elem) item.characterId = BsonUtils::toInt32(elem);
            if (auto elem = doc["name"]; elem) item.name = BsonUtils::toQString(elem);
            if (auto elem = doc["name_cn"]; elem) item.nameCn = BsonUtils::toQString(elem);
            if (auto elem = doc["role_type"]; elem) item.roleType = BsonUtils::toInt32(elem);
            if (auto elem = doc["order"]; elem) item.order = BsonUtils::toInt32(elem);
            if (auto elem = doc["url"]; elem) item.url = BsonUtils::toQString(elem);
            list.push_back(item);
        }
        return list;
    } catch (const std::exception &e) {
        return std::unexpected(QStringLiteral("getSubjectCharacters failed: ") + QString::fromUtf8(e.what()));
    }
}

DbResult<QList<BangumiEpisodeDoc> > BangumiRepository::getSubjectEpisodes(qint32 subjectId) {
    try {
        if (!MongoConnection::instance().isConnected()) return QList<BangumiEpisodeDoc>{};
        auto db = MongoConnection::instance().database("bangumi");
        auto coll = db["episodes"];

        mongocxx::options::find opts;
        opts.sort(make_document(kvp("sort", 1)));
        auto filter = make_document(kvp("subject_id", subjectId));
        auto cursor = coll.find(filter.view(), opts);
        QList<BangumiEpisodeDoc> list;
        for (auto &&doc: cursor) {
            BangumiEpisodeDoc e;
            if (auto elem = doc["_id"]; elem) e.id = BsonUtils::toInt32(elem);
            if (auto elem = doc["subject_id"]; elem) e.subjectId = BsonUtils::toInt32(elem);
            if (auto elem = doc["type"]; elem) e.type = BsonUtils::toInt32(elem);
            if (auto elem = doc["name"]; elem) e.name = BsonUtils::toQString(elem);
            if (auto elem = doc["name_cn"]; elem) e.nameCn = BsonUtils::toQString(elem);
            if (auto elem = doc["sort"]; elem) e.sort = BsonUtils::toInt32(elem);
            if (auto elem = doc["airdate"]; elem) e.airdate = BsonUtils::toQString(elem);
            if (auto elem = doc["duration"]; elem) e.duration = BsonUtils::toQString(elem);
            if (auto elem = doc["description"]; elem) e.description = BsonUtils::toQString(elem);
            if (auto elem = doc["disc"]; elem) e.disc = BsonUtils::toInt32(elem);
            list.push_back(e);
        }
        return list;
    } catch (const std::exception &e) {
        return std::unexpected(QStringLiteral("getSubjectEpisodes failed: ") + QString::fromUtf8(e.what()));
    }
}

DbResult<QList<BangumiSubjectRelationItem> > BangumiRepository::getSubjectRelations(qint32 subjectId) {
    try {
        if (!MongoConnection::instance().isConnected()) return QList<BangumiSubjectRelationItem>{};
        auto db = MongoConnection::instance().database("bangumi");
        auto coll = db["subject_relations"];

        mongocxx::pipeline pipeline;
        pipeline.match(make_document(kvp("subject_id", subjectId)));
        pipeline.lookup(make_document(
            kvp("from", "subjects"),
            kvp("localField", "related_subject_id"),
            kvp("foreignField", "_id"),
            kvp("as", "related_subject")
        ));
        pipeline.unwind(make_document(kvp("path", "$related_subject")));
        pipeline.project(make_document(
            kvp("subject_id", "$related_subject._id"),
            kvp("name", "$related_subject.name"),
            kvp("name_cn", "$related_subject.name_cn"),
            kvp("relation_type", "$relation_info.cn"),
            kvp("url", make_document(kvp("$concat", make_array(
                                             QStringLiteral("https://bgm.tv/subject/").toStdString(),
                                             make_document(kvp("$toString", "$related_subject._id"))
                                         ))))
        ));

        auto cursor = coll.aggregate(pipeline);
        QList<BangumiSubjectRelationItem> list;
        for (auto &&doc: cursor) {
            BangumiSubjectRelationItem item;
            if (auto elem = doc["subject_id"]; elem) item.subjectId = BsonUtils::toInt32(elem);
            if (auto elem = doc["name"]; elem) item.name = BsonUtils::toQString(elem);
            if (auto elem = doc["name_cn"]; elem) item.nameCn = BsonUtils::toQString(elem);
            if (auto elem = doc["relation_type"]; elem) item.relationType = BsonUtils::toQString(elem);
            if (auto elem = doc["url"]; elem) item.url = BsonUtils::toQString(elem);
            list.push_back(item);
        }
        return list;
    } catch (const std::exception &e) {
        return std::unexpected(QStringLiteral("getSubjectRelations failed: ") + QString::fromUtf8(e.what()));
    }
}

DbResult<BangumiSubjectDetail> BangumiRepository::getSubjectDetail(qint32 subjectId) {
    try {
        BangumiSubjectDetail detail;
        auto subject = getSubjectById(subjectId);
        if (!subject) return std::unexpected(subject.error());
        if (subject->id == 0) return detail;
        static_cast<BangumiSubjectDoc &>(detail) = *subject;

        auto staff = getSubjectStaff(subjectId);
        if (!staff) return std::unexpected(staff.error());
        detail.staff = std::move(*staff);

        auto chars = getSubjectCharacters(subjectId);
        if (!chars) return std::unexpected(chars.error());
        detail.characters = std::move(*chars);

        auto eps = getSubjectEpisodes(subjectId);
        if (!eps) return std::unexpected(eps.error());
        detail.episodes = std::move(*eps);

        auto rels = getSubjectRelations(subjectId);
        if (!rels) return std::unexpected(rels.error());
        detail.relations = std::move(*rels);

        return detail;
    } catch (const std::exception &e) {
        return std::unexpected(QStringLiteral("getSubjectDetail failed: ") + QString::fromUtf8(e.what()));
    }
}

DbResult<QList<BangumiSubjectDoc> > BangumiRepository::getAllSubjects(qint32 batchSize, qint32 skip) {
    try {
        if (!MongoConnection::instance().isConnected()) return QList<BangumiSubjectDoc>{};
        auto db = MongoConnection::instance().database("bangumi");
        auto coll = db["subjects"];

        mongocxx::options::find opts;
        opts.skip(skip);
        opts.limit(batchSize);
        auto cursor = coll.find({}, opts);
        QList<BangumiSubjectDoc> results;
        for (auto &&doc: cursor) {
            results.push_back(parseSubjectDoc(doc));
        }
        return results;
    } catch (const std::exception &e) {
        return std::unexpected(QStringLiteral("getAllSubjects failed: ") + QString::fromUtf8(e.what()));
    }
}

DbResult<qint32> BangumiRepository::getTotalSubjectsCount() {
    try {
        if (!MongoConnection::instance().isConnected()) return 0;
        auto db = MongoConnection::instance().database("bangumi");
        auto coll = db["subjects"];
        return static_cast<qint32>(coll.count_documents({}));
    } catch (const std::exception &e) {
        return std::unexpected(QStringLiteral("getTotalSubjectsCount failed: ") + QString::fromUtf8(e.what()));
    }
}
